lang/java/reef-utils/src/test/java/org/apache/reef/util/MultiAsyncToSyncTest.java

```diff
@@ -0,0 +1,225 @@
+      }
+    } catch (Exception e) {
+      LOG.log(Level.INFO, "Caught exception waiting for completion...", e);
+    }
+  }
+}
+
+/**
+ * Verify proper operation of the MultiAsyncToSync class.
+ */
+public final class MultiAsyncToSyncTest {
+  private static final Logger LOG = Logger.getLogger(MultiAsyncToSyncTest.class.getName());
+
+  /**
+   * Verify calculations successfully complete when no timeout occurs.
+   */
+  @Test
+  public void testNoTimeout() throws InterruptedException, ExecutionException {
+    LOG.log(Level.INFO, "Starting...");
+
+    // Parameters that do not force a timeout.
+    final int incrementerSleepTimeSeconds = 2;
+    final long timeoutPeriodSeconds = 4;
+    final int input = 1;
+
+    int result = 0;
+    SynchronousApi apiObject = new SynchronousApi(incrementerSleepTimeSeconds, timeoutPeriodSeconds);
+    result = apiObject.apiCall(input);
+    apiObject.complete();
+    Assert.assertTrue("Value incremented by one", result == (input + 1));
+  }
+
+  /**
+   * Verify an error is returned when a timeout occurs.
+   */
+  @Test
+  public void testTimeout() throws
+      InvalidBlockedCallerIdentifierException, InterruptedException, ExecutionException {
+    LOG.log(Level.INFO, "Starting...");
+
+    // Parameters that do not force a timeout.
+    final int incrementerSleepTimeSeconds = 4;
+    final long timeoutPeriodSeconds = 2;
+    final int input = 1;
+
+    int result = 0;
+    SynchronousApi apiObject = new SynchronousApi(incrementerSleepTimeSeconds, timeoutPeriodSeconds);
+    result = apiObject.apiCall(input);
+    apiObject.complete();
+    Assert.assertTrue("Timeout occurred", result == 0);
+  }
+
+  /**
+   * Verify no interaction occurs when multiple calls are in flight.
+   */
+  @Test
+  public void testMultipleCalls() throws InterruptedException,
+    ExecutionException, InvocationTargetException, IllegalAccessException, NoSuchMethodException {
+
+    LOG.log(Level.INFO, "Starting...");
+
+    // Parameters that do not force a timeout.
+    final int incrementerSleepTimeSeconds = 2;
+    final long timeoutPeriodSeconds = 4;
+    final SynchronousApi apiObject =
+        new SynchronousApi(incrementerSleepTimeSeconds, timeoutPeriodSeconds);
+
+    final String function = "apiCall";
+    final int input = 1;
+    final FutureTask<Integer> task1 =
+        new FutureTask<>(new MethodCallable<Integer>(apiObject, function, input));
+    final FutureTask<Integer> task2
+        = new FutureTask<>(new MethodCallable<Integer>(apiObject, function, input + 1));
+
+    // Execute API calls concurrently.
+    final ExecutorService executor = Executors.newFixedThreadPool(2);
+    executor.execute(task1);
+    executor.execute(task2);
+
+    final int result1 = task1.get();
+    final int result2 = task2.get();
+
+    apiObject.complete();
+
+    Assert.assertTrue("Input incremented by one", result1 == (input + 1));
+    Assert.assertTrue("Input incremented by one", result2 == (input + 2));
+  }
+}
```

Comment:
Also, can we have a test to check for race conditions? Something along the lines of
```java
@Test
public void testMultipleCalls2() throws Exception {

  // Parameters that do not force a timeout.
  final int incrementerSleepTimeSeconds = 1;
  final long timeoutPeriodSeconds = 2;
  final String function = "apiCall";

  final int nTasks = 100;
  final FutureTask[] tasks = new FutureTask[nTasks];

  final ExecutorService executor = Executors.newFixedThreadPool(10);

  try (final SynchronousApi apiObject = new SynchronousApi(incrementerSleepTimeSeconds, timeoutPeriodSeconds)) {

    for (int i = 0; i < nTasks; ++i) {
      final FutureTask<Integer> task = new FutureTask<>(new MethodCallable<Integer>(apiObject, function, i));
      executor.execute(task);
      tasks[i] = task;
    }

    for (int i = 0; i < nTasks; ++i) {
      final int result = (Integer) tasks[i].get();
      Assert.assertEquals("Input must be incremented by one", i + 1, result);
    }
  }
}
```
(assuming that `SimpleApi` implements `AutoCloseable`)